src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -0,0 +1,288 @@
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
+{
+    internal class SemanticTokensHelpers
+    {
+        /// <summary>
+        /// Returns the semantic tokens for a given document with an optional range.
+        /// </summary>
+        internal static async Task<LSP.SemanticTokens> ComputeSemanticTokensAsync(
+            TextDocumentIdentifier? textDocument,
+            int previousResultId,
+            string? clientName,
+            ILspSolutionProvider solutionProvider,
+            LSP.Range? range,
+            CancellationToken cancellationToken)
+        {
+            if (textDocument == null)
+            {
+                return new LSP.SemanticTokens();
+            }
+
+            var document = solutionProvider.GetDocument(textDocument, clientName);
+            if (document == null)
+            {
+                return new LSP.SemanticTokens();
+            }
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            if (root == null)
+            {
+                return new LSP.SemanticTokens();
+            }
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            if (text == null)
+            {
+                return new LSP.SemanticTokens();
+            }
+
+            var textSpan = root.FullSpan;
+            if (range != null)
+            {
+                textSpan = ProtocolConversions.RangeToTextSpan(range, text);
+            }
+
+            var classifiedSpans = await Classifier.GetClassifiedSpansAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (classifiedSpans == null)
+            {
+                return new LSP.SemanticTokens();
+            }
+
+            var lastLineNumber = 0;
+            var lastStartCharacter = 0;
+
+            var groupedSpans = classifiedSpans.GroupBy(s => s.TextSpan);
+
+            using var _ = ArrayBuilder<int>.GetInstance(out var data);
+
+            // TO-DO: We should implement support for streaming once this LSP bug is fixed:
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1132601
+            var tokens = ComputeTokens(
+                data, ref lastLineNumber, ref lastStartCharacter, text.Lines, groupedSpans);
+
+            return new LSP.SemanticTokens { ResultId = (previousResultId + 1).ToString(), Data = tokens };
+        }
+
+        private static int[] ComputeTokens(
+            ArrayBuilder<int> data,
+            ref int lastLineNumber,
+            ref int lastStartCharacter,
+            TextLineCollection lines,
+            IEnumerable<IGrouping<TextSpan, ClassifiedSpan>> groupedSpans)
+        {
+            foreach (var span in groupedSpans)
+            {
+                ComputeNextToken(data, lines, ref lastLineNumber, ref lastStartCharacter, span);
+            }
+
+            return data.ToArray();
```

Comment:
this sees unnecessary.  caller passes in the builder.  tehy can do the conversion to an array.